### PR TITLE
[IOTDB-3711] Update RegionRouteMap based on heartbeat sample

### PR DESCRIPTION
**Pull request by CRZbulabula**
Update the RegionRouteMap cache in PartitionCache based on the heartbeat sample. Update algorithms are distinguished according to different routing policies:

1. For greedy policy, update RegionRouteMap when there exists some DataNodes fail down.
2. For leader policy, update RegionRouteMap when the leadership of Regions changed.

We'll see log like this when a DataNode fails down:

`2022-07-05 23:48:21,104 [pool-3-IoTDB-LoadManager-1] INFO  o.a.i.c.m.l.LoadManager:302 - Begin to broadcast RegionRouteMap: {TConsensusGroupId(type:DataRegion, id:1)=TRegionReplicaSet(regionId:TConsensusGroupId(type:DataRegion, id:1), dataNodeLocations:[TDataNodeLocation(dataNodeId:2, clientRpcEndPoint:TEndPoint(ip:127.0.0.1, port:6668), internalEndPoint:TEndPoint(ip:127.0.0.1, port:9004), mPPDataExchangeEndPoint:TEndPoint(ip:127.0.0.1, port:8778), dataRegionConsensusEndPoint:TEndPoint(ip:127.0.0.1, port:40011), schemaRegionConsensusEndPoint:TEndPoint(ip:127.0.0.1, port:50011))]), TConsensusGroupId(type:SchemaRegion, id:0)=TRegionReplicaSet(regionId:TConsensusGroupId(type:SchemaRegion, id:0), dataNodeLocations:[TDataNodeLocation(dataNodeId:3, clientRpcEndPoint:TEndPoint(ip:127.0.0.1, port:6669), internalEndPoint:TEndPoint(ip:127.0.0.1, port:9005), mPPDataExchangeEndPoint:TEndPoint(ip:127.0.0.1, port:8779), dataRegionConsensusEndPoint:TEndPoint(ip:127.0.0.1, port:40012), schemaRegionConsensusEndPoint:TEndPoint(ip:127.0.0.1, port:50012))])} `

### CODE REVIEW

**Review by SpriCoder — APPROVED** [review]
LGTM~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/DataNodeHeartbeatCache.java [hunk 1]
```diff
@@ -44,13 +44,13 @@ public DataNodeHeartbeatCache() {
...
+      // Only sequential HeartbeatSamples are accepted.
+      // And un-sequential HeartbeatSamples will be discarded.
       if (slidingWindow.size() == 0
-          || slidingWindow.getLast().getSendTimestamp() < newHeartbeat.getSendTimestamp()) {
-        slidingWindow.add(newHeartbeat);
+          || slidingWindow.getLast().getSendTimestamp() < newHeartbeatSample.getSendTimestamp()) {
+        slidingWindow.add(newHeartbeatSample);
       }
 
       while (slidingWindow.size() > maximumWindowSize) {
```
Comment (xingtanzjr) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/DataNodeHeartbeatCache.java:56]: Be confident, using `if` is enough here.

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/ConfigNodeHeartbeatCache.java [hunk 2]
```diff
@@ -18,33 +18,37 @@
...
       // Only sequential heartbeats are accepted.
       // And un-sequential heartbeats will be discarded.
       if (slidingWindow.size() == 0
-          || slidingWindow.getLast().getSendTimestamp() < newHeartbeat.getSendTimestamp()) {
-        slidingWindow.add(newHeartbeat);
+          || slidingWindow.getLast().getSendTimestamp() < newHeartbeatSample.getSendTimestamp()) {
+        slidingWindow.add(newHeartbeatSample);
       }
 
       while (slidingWindow.size() > maximumWindowSize) {
```
Comment (xingtanzjr) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/ConfigNodeHeartbeatCache.java:54]: `if` is enough here

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/RegionGroupCache.java [hunk 3]
```diff
@@ -18,26 +18,75 @@
...
+            updateVersion = lastSample.getSendTimestamp();
+            updateLeaderDataNodeId = lastSample.getBelongedDataNodeId();
+          }
+        }
+      }
+    }
+
+    if (updateVersion > versionTimestamp.get()) {
+      versionTimestamp.set(updateVersion);
+      leaderDataNodeId.set(updateLeaderDataNodeId);
```
[the remark this answers is not in the document]
Comment (xingtanzjr) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/RegionGroupCache.java:86]: False alarm

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/ConfigNodeHeartbeatCache.java (hunk 2 above)
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/ConfigNodeHeartbeatCache.java]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/RegionGroupCache.java [hunk 4]
```diff
@@ -18,26 +18,75 @@
...
+          slidingWindow.get(newHeartbeatSample.getBelongedDataNodeId());
+
+      // Only sequential HeartbeatSamples are accepted.
+      // And un-sequential HeartbeatSamples will be discarded.
+      if (samples.size() == 0
+          || samples.getLast().getSendTimestamp() < newHeartbeatSample.getSendTimestamp()) {
+        samples.add(newHeartbeatSample);
+      }
+
+      while (samples.size() > maximumWindowSize) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/heartbeat/RegionGroupCache.java]: fixed~

**Pull request merged**